gusto/state.py

```diff
@@ -229,31 +295,26 @@ def dump(self, t=0, diagnostic_everydump=False, pickup=False):
                 t = chk.read_attribute("/", "time")
                 next(self.dumpcount)
 
-        elif (next(self.dumpcount) % self.output.dumpfreq) == 0:
-
-            print("DBG dumping", t)
+        else:
 
             # calculate diagnostic fields
             for field in self.diagnostic_fields:
                 field(self)
 
-            # dump fields
-            self.dumpfile.write(*self.to_dump)
-
-            # dump fields on latlon mesh
-            if len(self.output.dumplist_latlon) > 0:
-                self.dumpfile_ll.write(*self.to_dump_latlon)
-
             # compute diagnostics
-            diagnostic_fns = ['min', 'max', 'rms', 'l2']
             for name in self.diagnostics.fields:
-                for fn in diagnostic_fns:
+                for fn in self.diagnostics.available_diagnostics:
                     d = getattr(self.diagnostics, fn)
                     data = d(self.fields(name))
-                    self.diagnostic_data[name][fn].append(data)
-                if len(self.fields(name).ufl_shape) == 0:
-                    data = self.diagnostics.total(self.fields(name))
-                    self.diagnostic_data[name]["total"].append(data)
+                    self.diagnostic_data[name][fn] = data
+            self.diagnostic_dump()
+
+            # calculate pointwise data
+            point_data = {}
+            for name, points in self.output.point_data:
+                # get points in the right format for the at function
+                point_data[name] = np.asarray(self.fields(name).at(points))
+            self.pointwise_dump(point_data)
```

Comment:
It seems that one could move the evaluation of the fields into the `pointwise_dump` function (similarly with the `diagnostic_dump`).  That way you have less "action at a distance".  I would also be tempted to pull these out as free functions.  Let me push a commit showing what I mean.